Chore/plugin update

JacquesDurand opened the pull request:
<!-- Please update this template with something that matches your PR -->
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | fixes #... <!-- prefix each issue number with "fixes #", if any -->


This PR is an attempt to continue the work done in #62 and also make the plugin compatible with Sylius 1.12 and Symfony 6

I also integrated a ```docker-compose.yml``` including a mysql database for development/contributing purposes.
(updated documentation about contributing is missing, I'll add this asap).

I am unsure about any breaking changes, but tests are passing locally and I did not change the original sources.

Any feedback is deeply appreciated and I hope this will help !

dunglas committed:
chore: compatibility with PHP 8

dunglas committed:
chore: php-cs-fixer

dunglas committed:
fix: cs

JulieNobody committed:
functional version PHP8.0

JulieNobody committed:
functional version PHP8.0 test ok

JulieNobody committed:
fix/misc

JulieNobody committed:
delete gulp-sass

JulieNobody committed:
update gulp-sass

JacquesDurand committed:
chore(deps): update plugin to be Symfony 6 + Sylius 1.12 compatible

JacquesDurand committed:
fix(flex): add allow-plugin in composer.json

JacquesDurand committed:
docs: mention docker-compose

JacquesDurand committed:
fix(ci): rerun workflows on synchronize

JacquesDurand committed:
fix(ci): symfony flex configuration

JacquesDurand committed:
fix(ci): revert database url

JacquesDurand committed:
chore: properly ignore drivers

JacquesDurand committed:
chore: properly ignore build files

JacquesDurand committed:
chore(deps): bump sylius calendar

JacquesDurand committed:
fix(controller): remove session injection

JacquesDurand committed:
fix(di): lock

JacquesDurand committed:
feat: use attributes instead of annotations

JacquesDurand committed:
chore(deps): bump minimal php + sylius requirements

JacquesDurand committed:
feat: use attributes in test app

JacquesDurand committed:
fix: new symfony session config in test app

JacquesDurand commented:
Note: the "fix" I provide for the issue with the Dependency Injection concerning the lock might totally be out of touch as I am not completely confident about locks and their usage in this case. Do not hesitate to tell me if any changes are needed.

dunglas commented on .github/workflows/ci.yml:
```diff
@@ -17,7 +22,9 @@ jobs:
-        run: composer global require -q --no-ansi --no-interaction --no-scripts --no-suggest --no-progress --prefer-dist symfony/flex
+        run: |
+          composer global config --no-plugins allow-plugins.symfony/flex true
+          composer global require --no-progress --no-scripts --no-plugins "symfony/flex:^1.10"
```
IIRC the feature provided by Flex to resolve dependencies faster has now been merged in Composer, and this can now be removed:

```suggestion
```

dunglas commented on CONTRIBUTING.md:
```diff
@@ -9,8 +9,9 @@ that is full of examples.
 ## Quickstart Installation
 
 1. Run `composer install`.
+2. Run `docker-compose up -d` to set up the local Mysql database
```
```suggestion
2. Run `docker compose up -d` to set up the local MySQL database
```

dunglas commented on composer.json:
```diff
@@ -5,28 +5,35 @@
+        "sylius/calendar": "^0.4.0",
+        "sylius/mailer-bundle": "^2.0",
+        "sylius/sylius": "^1.11",
+        "symfony/lock": "^4.4 || ^5.0 || ^6.0",
```
```suggestion
        "symfony/lock": "^5.4 || ^6.0",
```

Symfony 4 is EOLed, I wouldn't bother supporting it. Also, we can bump to 5.4 (older versions are EOLed). Same for other Symfony packages.

dunglas commented on composer.json:
```diff
@@ -5,28 +5,35 @@
+        "symfony/webpack-encore-bundle": "^1.16"
     },
     "require-dev": {
+        "dbrekelmans/bdi": "^1.0",
```
This package is probably not necessary.

JacquesDurand replied:
It seems necessary to run `tests/E2E/ShopTest` with panther ( to get the necessary chrome/gecko drivers)

dunglas replied:
Chromedriver / Geckodriver could also be installed using the Package manager of the distribution, and they are already included in GitHub Actions default images.

dunglas commented on docker-compose.yaml:
```diff
@@ -0,0 +1,8 @@
+    environment:
+      MYSQL_ROOT_PASSWORD: mysql
+    ports:
+      - ${MYSQL_PORT:-3306}:3306
```
```suggestion
      - ${MYSQL_PORT:-3306}:3306

```

dunglas commented on docker-compose.yaml:
```diff
@@ -0,0 +1,8 @@
+services:
+  mysql:
+    image: mysql:8.0
+    platform: linux/amd64
```
Is this necessary? ARM images are available for MySQL.

dunglas commented on src/Entity/ClickNCollectShipment.php:
```diff
@@ -21,20 +21,14 @@
-    /**
-     * @ORM\ManyToOne(targetEntity=\CoopTilleuls\SyliusClickNCollectPlugin\Entity\LocationInterface::class)
-     */
+    #[ORM\ManyToOne(targetEntity:\CoopTilleuls\SyliusClickNCollectPlugin\Entity\LocationInterface::class )]
```
```suggestion
    #[ORM\ManyToOne(targetEntity: \CoopTilleuls\SyliusClickNCollectPlugin\Entity\LocationInterface::class )]
```

Also, you can add a `use` statement instead of using the FQCN.
With recent versions of Doctrine, this can even be removed (this will be guessed using the type);

dunglas commented on src/Entity/ClickNCollectShippingMethod.php:
```diff
@@ -22,16 +22,10 @@
-     * )
-     */
-    protected $locations;
+    #[ORM\ManyToMany(targetEntity:\CoopTilleuls\SyliusClickNCollectPlugin\Entity\LocationInterface::class, inversedBy: "shippingMethods" )]
```
Sae here regarding the CS.

dunglas commented on src/Entity/ClickNCollectShippingMethod.php:
```diff
@@ -22,16 +22,10 @@
 trait ClickNCollectShippingMethod
 {
-    /**
-     * @var Collection|LocationInterface[]
```
You should keep this for auto-completion, static analysis, etc. This can be: `Collection<LocationInterface>`

dunglas commented on src/EventListener/ShipmentCollectionTimeLockListener.php:
```diff
@@ -34,11 +37,12 @@ final class ShipmentCollectionTimeLockListener
     private string $shipmentClass;
 
-    public function __construct(EntityManagerInterface $entityManager, LockInterface $lock, CollectionTimeRepositoryInterface $collectionTimeRepository)
+    public function __construct(EntityManagerInterface $entityManager,  CollectionTimeRepositoryInterface $collectionTimeRepository)
```
You should inject the `LockFactory` service instead and use it: https://symfony.com/doc/current/lock.html

dunglas commented on src/Validator/Constraints/Rrule.php:
```diff
@@ -19,9 +19,8 @@
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
- *
- * @Annotation
```
You should keep this for backward compatibility with older versions.

dunglas commented on tests/Application/.env:
```diff
@@ -12,12 +12,9 @@ APP_SECRET=EDITME
-# Delivery is disabled by default via "null://localhost"
-MAILER_URL=smtp://localhost
-###< symfony/swiftmailer-bundle ###
+MAILER_DSN=nul://null
```
Flex markup is missing, shouldn't this variables be added/removed by Flex?

JacquesDurand replied:
Flex was installed globally in the CI, should I add it and run recipes ?

dunglas commented on tests/Application/Kernel.php:
```diff
@@ -44,75 +31,79 @@ final class Kernel extends BaseKernel
     protected function getContainerBaseClass(): string
     {
-        if ($this->isTestEnvironment()) {
+        if ($this->isTestEnvironment() && class_exists(MockerContainer::class)) {
```
Shouldn't this class always exist with the test env?

dunglas commented on tests/Application/Kernel.php:
```diff
@@ -44,75 +31,79 @@ final class Kernel extends BaseKernel
-        ]);
-
-        return new DelegatingLoader($resolver);
+        return 0 === strpos($this->getEnvironment(), 'test');
```
```suggestion
        return str_starts_with($this->getEnvironment(), 'test');
```

IMHO we could even remove entirely this private method as it's used only one time.

dunglas commented on tests/Application/config/bootstrap.php:
```diff
@@ -1,14 +1,5 @@
 <?php
 
-/*
```
This header should be kept.

dunglas commented on tests/Application/config/packages/dev/mailer.yaml:
```diff
@@ -0,0 +1,3 @@
+framework:
+  mailer:
+    dsn: 'null://null'
```
```suggestion
    dsn: 'null://null'

```

dunglas commented on tests/Application/config/packages/test/mailer.yaml:
```diff
@@ -0,0 +1,3 @@
+framework:
+  mailer:
+    dsn: 'null://null'
```
```suggestion
    dsn: 'null://null'

```

dunglas commented on tests/Application/public/index.php:
```diff
@@ -1,21 +1,12 @@
 <?php
 
-/*
```
same here

dunglas reviewed (changes requested):
Great work! Thank you.

I made some minor comments. Could you also run PHP CS Fixer, please? It should fix most of the CS issues (or its config needs to be improved).

JacquesDurand committed:
fix: CS + review

louis-iteracode committed:
👽️ fix - update rrule script path

Maintainer of the lib rrule has changed his pseudo from jakubroztocil to jkbrzt

JacquesDurand committed:
fix(config): require pagerfanta

romainallanot commented:
closed this as merged in #68
